core/src/main/scala/chisel3/VerificationStatement.scala

```diff
@@ -0,0 +1,236 @@
+  def apply(cond: Bool, message: String, data: Bits*)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Assume = macro _applyMacroWithMessage
+  def apply(cond: Bool)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Assume = macro _applyMacroWithNoMessage
+
+  /** An elaboration-time assumption. Calls the built-in Scala assume function. */
+  def apply(cond: Boolean, message: => String): Unit = Predef.assume(cond, message)
+  /** An elaboration-time assumption. Calls the built-in Scala assume function. */
+  def apply(cond: Boolean): Unit = Predef.assume(cond, "")
+
+  /** Named class for assumptions. */
+  final class Assume private[chisel3]() extends VerificationStatement
+
+  import VerificationStatement._
+
+  def _applyMacroWithMessage(c: blackbox.Context)(cond: c.Tree, message: c.Tree, data: c.Tree*)(sourceInfo: c.Tree, compileOptions: c.Tree): c.Tree = {
+    import c.universe._
+    val apply_impl_do = symbolOf[this.type].asClass.module.info.member(TermName("_applyWithSourceLine"))
+    q"$apply_impl_do($cond, ${getLine(c)}, _root_.scala.Some($message), ..$data)($sourceInfo, $compileOptions)"
+  }
+
+  def _applyMacroWithNoMessage(c: blackbox.Context)(cond: c.Tree)(sourceInfo: c.Tree, compileOptions: c.Tree): c.Tree = {
+    import c.universe._
+    val apply_impl_do = symbolOf[this.type].asClass.module.info.member(TermName("_applyWithSourceLine"))
+    q"$apply_impl_do($cond, ${getLine(c)}, _root_.scala.None)($sourceInfo, $compileOptions)"
+  }
+
+  /** Used by our macros. Do not call directly! */
+  def _applyWithSourceLine(cond: Bool, line: SourceLineInfo, message: Option[String], data: Bits*)
+    (implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Assume = {
+    val id = new Assume()
+    when(!Module.reset.asBool()) {
+      Builder.pushCommand(Verification(id, Formal.Assume, sourceInfo, Module.clock.ref, cond.ref, ""))
+      failureMessage("Assumption", line, cond, message, data)
+    }
+    id
+  }
+}
+
+
+object cover {
+  /** Declares a condition to be covered.
+    * At ever clock event, a counter is incremented iff the condition is active
+    * and reset is inactive.
+    *
+    * Does not fire when in reset (defined as the encapsulating Module's
+    * reset). If your definition of reset is not the encapsulating Module's
+    * reset, you will need to gate this externally.
+    *
+    * May be called outside of a Module (like defined in a function), so
+    * functions using assert make the standard Module assumptions (single clock
+    * and single reset).
+    *
+    * @param cond condition that will be sampled on every clock tick
+    * @param message a string describing the cover event
+    */
+  // Macros currently can't take default arguments, so we need two functions to emulate defaults.
+  def apply(cond: Bool, message: String)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Cover = macro _applyMacroWithMessage
```

Comment:
why does this one not also take a `Bits*`?

Reply:
I did not find a reasonable way to use a `printf` statement with a `cover` statement.
The failure case with a `cover` is if it cannot be covered. In that case, there are no values to print.
If you have an idea on how a `printf` statement could make sense with cover, I would be happy to hear about it.

Reply:
ah ok never mind, i see what you mean now!